### PR TITLE
Improve vscode tab API

### PR DESCRIPTION
**Pull request by jonah-iden**
#### What it does

This improves and fixes some things for the vscode tab api.
- Now there should always be a default tabGroup open, even when no tabs are open
- Implements the viewCollumn property
- Smaller fix for tab events being fired even though there was no change

#### How to test

Here is a small test extension.
This shows message and logs into the console the different tabApi events.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by tsmaeder** [comment]
@jonah-iden do we have a list of issues this PR fixes?

**Comment by jonah-iden** [comment]
It can easily be that i overlooked problems. Its sadly super hard to align this perfectly to vscode because phosphor is just doing things differently to vscodes layouting. Like events just happening in different order

**Comment by tsmaeder** [comment]
https://github.com/eclipse-theia/theia/issues/12702
https://github.com/eclipse-theia/theia/issues/13674

Also, we should try to remove the workaround in the tests I introduced for the 1.88.1 built-ins. https://github.com/eclipse-theia/theia/issues/13679#issuecomment-2094653160

**Comment by tsmaeder** [comment]
Removing the workaround for https://github.com/eclipse-theia/theia/issues/13679 makes the tests fail again. Seems that one at least is still there.

File: packages/plugin-ext/src/main/browser/tabs/tabs-main.ts [hunk 1]
```diff
@@ -99,16 +109,22 @@ export class TabsMainImpl implements TabsMain, Disposable {
...
         this.tabInfoLookup.clear();
         this.disposableTabBarListeners.dispose();
-        this.applicationShell.mainAreaTabBars.forEach(tabBar => {
-            this.attachListenersToTabBar(tabBar);
-            const groupDto = this.createTabGroupDto(tabBar);
-            tabBar.titles.forEach((title, index) => this.tabInfoLookup.set(title, { group: groupDto, tab: groupDto.tabs[index], tabIndex: index }));
-            newTabGroupModel.set(tabBar, groupDto);
-        });
+        this.applicationShell.mainAreaTabBars
+            .concat(newTabbar ? [newTabbar] : [])
```
[the remark this answers is not in the document]
Comment (jonah-iden) [reply packages/plugin-ext/src/main/browser/tabs/tabs-main.ts:121]: yeah thats an event ordering issue. There is sadly no way we can hook in to an event after the toolbar is added but before the widget is, so we need to add this here

File: examples/api-tests/src/typescript.spec.js [hunk 2]
```diff
@@ -103,7 +103,6 @@ describe('TypeScript', function () {
         const editorWidget = widget instanceof EditorWidget ? widget : undefined;
         const editor = MonacoEditor.get(editorWidget);
         assert.isDefined(editor);
-        await timeout(1000); // workaround for https://github.com/eclipse-theia/theia/issues/13679
```
[the remark this answers is not in the document]
Comment (tsmaeder) [reply examples/api-tests/src/typescript.spec.js:106]: Also, I need the 1 second timeout for the tests to run reliably, locally.

**Review by tsmaeder — CHANGES_REQUESTED** [review]
Otherwise, looks good to me.

File: examples/api-tests/src/typescript.spec.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (msujew) [reply examples/api-tests/src/typescript.spec.js:106]: The `typescript.isManagedFile` context key is working unreliably. When opening a TypeScript file in the editor and querying the context key value, I always get `false`, even though a TypeScript editor is open (and focused).

I've reinstated the 1 second timeout, as the issue seems to be unrelated to the tab API.

**Comment by msujew** [comment]
@tsmaeder FYI the latest commit fixes an error in the plugin host that appeared due to an off-by-one error. Note that the `typescript.isManagedFile` context key seems to work unreliably and cannot be used to check whether the TypeScript language services are available (or not). See also https://github.com/eclipse-theia/theia/pull/13730#discussion_r1619148021. This could be potentially fixed in a separate PR. However it fixes the initial issue that causes https://github.com/eclipse-theia/theia/issues/13674, to this PR should be good to go.

File: examples/api-tests/src/typescript.spec.js [hunk 3]
```diff
@@ -103,7 +103,7 @@ describe('TypeScript', function () {
...
+        await timeout(1000);
```
Comment (tsmaeder) [inline examples/api-tests/src/typescript.spec.js:106]:
```suggestion
        await timeout(1000);// workaround for https://github.com/eclipse-theia/theia/issues/13679
```

Comment (tsmaeder) [reply examples/api-tests/src/typescript.spec.js:106]: We should keep the reference in order to mark the workaround

**Comment by msujew** [comment]
FYI I believe I've mostly fixed https://github.com/eclipse-theia/theia/issues/13679, it's just a minor issue left: The `onDidChangeActiveTextEditor` event is fired on the plugin host before the tab API has been updated on the plugin host (due to the order of events in the frontend). However, I couldn't find a quick way to fix this without breaking a bunch of other stuff, so I'm putting this off for a separate PR.

The original off-by-one issue mentioned in the issue is fixed, it's just a matter of correctly ordering the events.

**Review by tsmaeder — APPROVED** [review]

**Pull request merged**